cmd/notify-mailer/main.go

```diff
@@ -0,0 +1,181 @@
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	netmail "net/mail"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/jmhodges/clock"
+	"gopkg.in/gorp.v1"
+
+	"github.com/letsencrypt/boulder/cmd"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/mail"
+	"github.com/letsencrypt/boulder/sa"
+)
+
+type mailer struct {
+	clk           clock.Clock
+	log           blog.Logger
+	dbMap         *gorp.DbMap
+	mailer        mail.Mailer
+	subject       string
+	emailTemplate string
+	destinations  []string
+	checkpoint    interval
+	sleepInterval time.Duration
+}
+
+type interval struct {
+	start int
+	end   int
+}
+
+func (i *interval) ok() error {
+	if i.start < 0 || i.end < 0 {
+		return fmt.Errorf(
+			"interval start (%d) and end (%d) must both be positive integers",
+			i.start, i.end)
+	}
+
+	if i.start > i.end && i.end != 0 {
+		return fmt.Errorf(
+			"interval start value (%d) is greater than end value (%d)",
+			i.start, i.end)
+	}
+
+	return nil
+}
+
+func (m *mailer) run() error {
+	// Make sure the checkpoint range is OK
+	if checkpointErr := m.checkpoint.ok(); checkpointErr != nil {
+		return checkpointErr
+	}
+	// Do not allow a start larger than the # of destinations
+	if m.checkpoint.start > len(m.destinations) {
+		return fmt.Errorf(
+			"interval start value (%d) is greater than number of destinations (%d)",
+			m.checkpoint.start,
+			len(m.destinations))
+	}
+	// Do not allow a negative sleep interval
+	if m.sleepInterval < 0 {
+		return fmt.Errorf(
+			"sleep interval (%d) is < 0", m.sleepInterval)
+	}
+	// If there is no endpoint specified, use the total # of destinations
+	if m.checkpoint.end == 0 {
+		m.checkpoint.end = len(m.destinations)
+	}
+	for i, dest := range m.destinations {
```

Comment:
To avoid the step skipping `if` below you could use `m.destinations[m.checkpoint.start:m.checkpoint.end]`.